### ISSUE TITLE
Optional 'types' argument of Sphinx.add_config_value() is not documented

### ISSUE THREAD

**Issue by brechtm** [sphinx-doc/sphinx#8054]
See https://www.sphinx-doc.org/en/master/extdev/appapi.html?highlight=add_config_value#sphinx.application.Sphinx.add_config_value

**Comment by danieleades** [comment]
Hi @tk0miya,
I'm still slightly confused by this parameter. What *types* can the `types` be? Is this restricted to an iterable of primitive types? Or do they need to be *callables*/ have constructors? Is it possible to extend these to validate custom types? Is it possible to use `typing` types, and if not, is support planned for these? Would it make sense to accept a `typing.Union` instead of an iterable?

**Comment by tk0miya** [comment]
Sorry, I can't answer it well. Please read the implementation of types checker.
https://github.com/sphinx-doc/sphinx/blob/3.x/sphinx/config.py#L407-L459

> is support planned for these?

Nothing planned and discussed. So any contributions are welcome :-)

**Comment by danieleades** [comment]
Yeah I can't say that the behaviour is *immediately* obvious from looking at that implementation.

Would be really useful for this to be documented somewhere.

Should this issue be reopened?

**Comment by tk0miya** [comment]
Absolutely. Reopened.
